akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala

```diff
@@ -0,0 +1,247 @@
+   */
+  implicit class CircuitBreakerAwareFuture(val future: Future[Any]) extends AnyVal {
+    def failForOpenCircuit(implicit executionContext: ExecutionContext): Future[Any] = failForOpenCircuitWith(new OpenCircuitException)
+
+    def failForOpenCircuitWith(throwing: ⇒ Throwable)(implicit executionContext: ExecutionContext): Future[Any] = {
+      future.flatMap {
+        _ match {
+          case CircuitOpenFailure(_) ⇒ Future.failed(throwing)
+          case result                ⇒ Future.successful(result)
+        }
+      }
+    }
+
+  }
+
+}
+
+object CircuitBreakerInternalEvents {
+  sealed trait CircuitBreakerInternalEvent
+  case object CallFailed extends CircuitBreakerInternalEvent
+  case object CallSucceeded extends CircuitBreakerInternalEvent
+}
+
+import CircuitBreakerActor._
+import CircuitBreakerInternalEvents._
+
+class CircuitBreakerActor(
+  target: ActorRef,
+  maxFailures: Int,
+  callTimeout: Timeout,
+  resetTimeout: Timeout,
+  circuitEventListener: Option[ActorRef],
+  failureDetector: Any ⇒ Boolean,
+  failureMap: CircuitOpenFailure ⇒ Any) extends Actor with ActorLogging with FSM[CircuitBreakerState, CircuitBreakerStateData] {
+
+  startWith(Closed, CircuitBreakerStateData(failureCount = 0))
+
+  def callSucceededHandling: StateFunction = {
+    case Event(CallSucceeded, state) ⇒
+      log.debug("Received call succeeded notification in state {} resetting counter", state)
+      goto(Closed) using CircuitBreakerStateData(failureCount = 0, firstHalfOpenMessageSent = false)
+  }
+
+  when(Closed) {
+    callSucceededHandling orElse {
+      case Event(TellOnly(message), _) ⇒
+        log.debug("CLOSED: Sending message {} without expecting any response", message)
+        target ! message
+        stay
+
+      case Event(CallFailed, state) ⇒
+        log.debug("Received call failed notification in state {} incrementing counter", state)
+        val newState = state.copy(failureCount = state.failureCount + 1)
+        if (newState.failureCount < maxFailures) {
+          stay using newState
+        } else {
+          goto(Open) using newState
+        }
+
+      case Event(message, state) ⇒
+        log.debug("CLOSED: Sending message {} expecting a response withing timeout {}", message, callTimeout)
+        val currentSender = sender()
+        forwardRequest(message, sender, state, log)
+        stay
+
+    }
+  }
+
+  when(Open, stateTimeout = resetTimeout.duration) {
+    callSucceededHandling orElse {
+      case Event(StateTimeout, state) ⇒
+        log.debug("Timeout expired for state OPEN, going to half open")
+        goto(HalfOpen) using state.copy(firstHalfOpenMessageSent = false)
+
+      case Event(CallFailed, state) ⇒
+        log.debug("OPEN: Call received a further call failed notification, probably from a previous timed out event, ignoring")
+        stay
+
+      case Event(openNotification @ CircuitOpenFailure(_), _) ⇒
+        log.error("Why did I send message {} to myself?", openNotification)
```

Comment:
Actually that code has been introduced when I was trying to fix a bug :-) I'll improve the logging with a proper message...